src/storage/nssdb/ci.rs

```diff
@@ -12,6 +15,130 @@ use crate::token::TokenFacilities;
 use crate::CSPRNG;
 use crate::{sizeof, void_ptr};
 
+use zeroize::Zeroize;
+
+const SHA256_LEN: usize = 32;
+const MAX_KEY_CACHE_SIZE: usize = 128;
+
+enum KeyOp {
+    Encryption,
+    Signature,
+}
+
+#[derive(Debug)]
+struct LockedKey<'a> {
+    id: [u8; SHA256_LEN],
+    l: RwLockReadGuard<'a, BTreeMap<[u8; SHA256_LEN], Object>>,
+}
+
+impl LockedKey<'_> {
+    fn get_id(&self) -> [u8; SHA256_LEN] {
+        self.id
+    }
+
+    pub fn get_key<'a>(&'a self) -> Option<&'a Object> {
+        self.l.get(&self.id)
+    }
+}
+
+#[derive(Debug)]
+pub struct KeysWithCaching {
+    enckey: Option<Vec<u8>>,
+    cache: RwLock<BTreeMap<[u8; SHA256_LEN], Object>>,
+}
+
+impl Default for KeysWithCaching {
+    fn default() -> KeysWithCaching {
+        KeysWithCaching {
+            enckey: None,
+            cache: RwLock::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl Drop for KeysWithCaching {
+    fn drop(&mut self) {
+        if let Some(ref mut key) = &mut self.enckey {
+            key.zeroize();
+        }
+    }
+}
+
+impl KeysWithCaching {
+    pub fn available(&self) -> bool {
+        self.enckey.is_some()
+    }
+
+    fn get_key(&self) -> Result<&[u8]> {
+        match &self.enckey {
+            Some(ref key) => Ok(key.as_slice()),
+            None => Err(CKR_USER_NOT_LOGGED_IN)?,
+        }
+    }
+
+    pub fn check_key(&self, check: &[u8]) -> bool {
+        match &self.enckey {
+            Some(ref key) => key.as_slice() == check,
+            None => false,
+        }
+    }
+
+    pub fn set_key(&mut self, key: Vec<u8>) {
+        if let Some(ref mut oldkey) = &mut self.enckey {
+            oldkey.zeroize();
+        }
+        self.enckey = Some(key);
+    }
+
+    pub fn unset_key(&mut self) {
+        if let Some(ref mut key) = &mut self.enckey {
+            key.zeroize();
+            self.enckey = None;
+        }
+    }
+
+    fn get_cached_key(&self, id: &[u8; SHA256_LEN]) -> Option<LockedKey> {
+        if self.enckey.is_none() {
+            /* access to the cache is available only if enckey is set.
+             * When unset it means the user logged off and no
+             * keys should be available */
+            return None;
+        }
+        let read_lock = match self.cache.read() {
+            Ok(r) => r,
+            Err(_) => return None,
+        };
+        Some(LockedKey {
+            id: *id,
+            l: read_lock,
+        })
+    }
+
+    fn set_cached_key(&self, id: &[u8; SHA256_LEN], key: Object) -> Result<()> {
+        match self.cache.write() {
+            Ok(mut w) => {
+                if w.len() > MAX_KEY_CACHE_SIZE {
+                    let _ = w.pop_last();
```

Comment:
This is basically dropping a random item from the BTreeMap as it is ordered. Is there some way to be able to drop the oldest item without too much complication, or this is the best approximation we can get?